### PR TITLE
expression: fix wrong check in builtinCastStringAsDecimalSig.evalDecimal

### LINKED ISSUE #16825: wrong check in builtinCastStringAsDecimalSig.evalDecimal
``` go
func (b *builtinCastStringAsDecimalSig) evalDecimal(row chunk.Row) (res *types.MyDecimal, isNull bool, err error) {
	// ...
	res = new(types.MyDecimal)
	sc := b.ctx.GetSessionVars().StmtCtx
	if !(b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && res.IsNegative()) {
	// ...
}
```

The result of res.IsNegative() will always be false. What we need to check here is val.IsNegative() acutually. But the type of `val` is string, thus `val` does not have an `IsNegative` function. We can refer the check of `builtinCastStringAsInt.evalInt`

### PR DESCRIPTION
**Pull request by ThySinner**
### What problem does this PR solve?

Issue Number: close #16825

Problem Summary:
`builtinCastStringAsDecimalSig.evalDecimal` checks if the decimal in the string is negative by checking a new `MyDecimal`, whose negative flag is always `false`.

### What is changed and how it works?

Check if the value is negative by simply checking the sign character in the decimal string.

### Related changes

- Need to cherry-pick to the release branch

### Check List

Tests

- Unit test

Side effects

### Release note 

- fix wrong check in builtinCastStringAsDecimalSig.evalDecimal

### CODE REVIEW

**Comment by ThySinner** [comment]
OK. Thanks for the hints.

**Comment by Reminiscent** [comment]
@XuHuaiyu  PTAL

**Comment by ThySinner** [comment]
@XuHuaiyu @Reminiscent  Unit-tests added

File: expression/builtin_cast_vec_test.go [hunk 1]
```diff
@@ -152,6 +155,58 @@ func (s *testEvaluatorSuite) TestVectorizedBuiltinCastFunc(c *C) {
...
+		sign = -1
+	} else {
+		sign = 1
+	}
+
+	input := chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeString)}, 1024)
+	for i := 0; i < 1024; i++ {
+		d := new(types.MyDecimal)
+		f := sign * rand.Float64() * 100000
+		if err := d.FromFloat64(-f); err != nil {
```
Comment (Reminiscent) [inline expression/builtin_cast_vec_test.go:201]: Why is `d.FromFloat64(-f)` instead of `d.FromFloat64(f)`?

Comment (ThySinner) [reply expression/builtin_cast_vec_test.go:201]: Oh, a stupid typo mistake. Thanks

Comment (ThySinner) [reply expression/builtin_cast_vec_test.go]: fixed

**Pull request merged**